src/ripple/consensus/Consensus.h

```diff
@@ -1787,47 +1802,89 @@ Consensus<Adaptor>::updateOurPositions(bool const share)
+        // But if no close time meet the criteria, make no changes.
+        //
+        // This is implemented slightly differently for validators vs
+        // non-validators. For non-validators, it is sufficient to merely
+        // count the close times from all peer positions to determine
+        // the best. Validators, however, risk putting the network into an
+        // impasse unless they are able to change their own position without
+        // first having counted it towards the close time totals.
+        //
+        // Here's how the impasse could occur:
+        // Imagine 5 validators. 3 have close time t1, and 2 have t2.
+        // As consensus time increases, the threshVote threshold also increases.
+        // Once threshVote exceeds 60%, no members of either set of validators
+        // will change their close times.
+        //
+        // Avoiding the impasse means that validators should not necessarily
+        // count their own close time towards the total until they know
+        // the most popular among their peers, and then change their vote
+        // to that of the most popular. In this case, the validators in set
+        // t2 from the example would have switched to close time t1 and ended
+        // the network impasse.
+        //
+        // Another wrinkle, however, is that too many position changes
+        // from validators also has a destabilizing affect. Consensus can
+        // take longer if peers have to keep re-calculating positions,
+        // and mistakes can be made if peers settle on the wrong position.
+        // Therefore, avoiding an impasse should also minimize the likelihood
+        // of gratuitous change of position.
+        //
+        // The solution for validators is to first track whether it's
+        // possible that the network is at an impasse based on how much
+        // time this current consensus round has taken. This is reflected
+        // in the "stuck" boolean. Once possibly stuck, validators do
+        // not use the threshVote threshold to decide whether to change their
+        // vote. Instead, they add their own close time vote in with
+        // those of their peers only after they discover the most popular,
+        // and the greatest of those if there is a tie. This solution avoids
+        // an impasse while minimizing gratuitous close time changes.
+        //
+        // Determining whether there is close time consensus is very simple
+        // in comparison: if votes for the best close time meet or exceed
+        // threshConsensus, then we have close time consensus. Otherwise, not.
+
+        // First, find the best close time with which to agree.
+        std::multimap<int, NetClock::time_point> ctVotesByCount;
         for (auto& [t, v] : closeTimeVotes)
+            ctVotesByCount.insert({v, t});
+        int maxVote = ctVotesByCount.rbegin()->first;
+        NetClock::time_point const& bestCtime = ctVotesByCount.rbegin()->second;
```

Comment:
I believe the following will be a faster way to compute `maxVote` and `bestCtime` because it avoids allocations:

```c++
    int maxVote = 0;
    NetClock::time_point bestCtime{};
    for (auto const& [t, v] : closeTimeVotes)
    {
        if (maxVote <= v)
        {
            maxVote = v;
            if (bestCtime < t)
                bestCtime = t;
        }
    }
```

Reply:
Fixed. I used your implementation. Initially, I liked what I did because the new container placed the best choice at the very end for easy access. But your implementation makes explicit the criteria-- (1) that which has the max votes, and if there's a tie (2) the latest.